index.js

```diff
@@ -363,10 +363,15 @@
 
   //  Type#fantasy-land/equals :: Type ~> Type -> Boolean
   _Type.prototype['fantasy-land/equals'] = function(other) {
-    return Z.equals(this.type, other.type) &&
-           Z.equals(this.name, other.name) &&
-           Z.equals(this.url, other.url) &&
-           Z.equals(Z.toString(this), Z.toString(other));
+    return (
+      Z.equals(this.type, other.type) &&
+      Z.equals(this.name, other.name) &&
+      Z.equals(this.url, other.url) &&
+      Z.equals(this.keys, other.keys) &&
+      this.keys.every(function(k) {
+        return Z.equals(this.types[k].type, other.types[k].type);
```

Comment:
I was worried that `this.types[n]` might lead back to a type we're currently processing in the stack, leading to infinite recursion. Based on your example I now see that this can't happen. The `this.types` refers only to the *type* parameters and **not** to the *constructor* parameters.